# Before you start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

## Hypothesis Testing

In [1]:
!pip install -U scipy

In [2]:
# import numpy and pandas
import pandas as pd
import numpy as np

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the San Francisco. We will start by loading the dataset and examining its contents. 

In [3]:
salaries = pd.read_csv('C:/Users/ameli/IRONHACK/WEEK_4/DAY_3/AFTERNOON_SESSION/lab-Hypothesis-Testing/data/Salaries.csv')

C:\Users\ameli\AppData\Local\Temp\ipykernel_32792\633549221.py:1: DtypeWarning: Columns (3,4,5,6,12) have mixed types. Specify dtype option on import or set low_memory=False.
  salaries = pd.read_csv('C:/Users/ameli/IRONHACK/WEEK_4/DAY_3/AFTERNOON_SESSION/lab-Hypothesis-Testing/data/Salaries.csv')


Examine the `salaries` dataset using the `head` function below.

In [4]:
salaries.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay OvertimePay   OtherPay Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.0  400184.25      NaN  567595.43         567595.43   
1  155966.02   245131.88  137811.38      NaN  538909.28         538909.28   
2  212739.13   106088.18    16452.6      NaN  335279.91         335279.91   
3    77916.0    56120.71   198306.9      NaN  332343.61         332343.61   
4   134401.6      9737.0  182234.59      NaN  326373.19         326373.19   

   Year  Notes         Agency Status  
0  2011    NaN  San Francisco    NaN  
1  2011    NaN  San Francisco    NaN  
2  2011    NaN  San Francisco    NaN  
3  2011    NaN  San Francisco    NaN  
4  2011    NaN  San Francisco    NaN

We see from looking at the `head` function that there is quite a bit of missing data. Get the amount of missing data in every column

In [5]:
salaries.isna().sum()

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                605
OvertimePay              0
OtherPay                 0
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Notes               148654
Agency                   0
Status              110535
dtype: int64

Get the shape of the dataframe

In [6]:
salaries.shape

(148654, 13)

Given output of the previous two cells, drop the corresponding column and compute again the amount of missing values.

In [7]:
salaries.drop('Notes', axis=1, inplace=True)

In [8]:
salaries.isna().sum()

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                605
OvertimePay              0
OtherPay                 0
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Agency                   0
Status              110535
dtype: int64

Check out what are the possible values of the column "Status".

In [9]:
salaries['Status'].unique()

array([nan, 'PT', 'FT'], dtype=object)

Drop any row with missing values in the "Status" column and compute again the number of missing values.

In [10]:
salaries.dropna(axis= 0, subset= 'Status', inplace= True)

In [11]:
salaries.isna().sum()

Id                  0
EmployeeName        0
JobTitle            0
BasePay             0
OvertimePay         0
OtherPay            0
Benefits            0
TotalPay            0
TotalPayBenefits    0
Year                0
Agency              0
Status              0
dtype: int64

Check out the types of each column and see if they make sense.

In [12]:
salaries.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 38119 entries, 110531 to 148653
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Id                38119 non-null  int64  
 1   EmployeeName      38119 non-null  object 
 2   JobTitle          38119 non-null  object 
 3   BasePay           38119 non-null  object 
 4   OvertimePay       38119 non-null  object 
 5   OtherPay          38119 non-null  object 
 6   Benefits          38119 non-null  object 
 7   TotalPay          38119 non-null  float64
 8   TotalPayBenefits  38119 non-null  float64
 9   Year              38119 non-null  int64  
 10  Agency            38119 non-null  object 
 11  Status            38119 non-null  object 
dtypes: float64(2), int64(2), object(8)
memory usage: 3.8+ MB


#### Comments on column types:

- BasePay, OvertimePay, OtherPay, and Benefits values should be float type.
- Year could be DateTime type, but as it only contains one value, we'll leave it as it is for the moment.

Do any type conversions and reset the index.

In [13]:
salaries = salaries.astype({'BasePay': 'float64', 'OvertimePay': 'float64', 
                 'OtherPay': 'float64', 'Benefits': 'float64'})
salaries = salaries.reset_index(drop=True)

In [14]:
salaries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38119 entries, 0 to 38118
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Id                38119 non-null  int64  
 1   EmployeeName      38119 non-null  object 
 2   JobTitle          38119 non-null  object 
 3   BasePay           38119 non-null  float64
 4   OvertimePay       38119 non-null  float64
 5   OtherPay          38119 non-null  float64
 6   Benefits          38119 non-null  float64
 7   TotalPay          38119 non-null  float64
 8   TotalPayBenefits  38119 non-null  float64
 9   Year              38119 non-null  int64  
 10  Agency            38119 non-null  object 
 11  Status            38119 non-null  object 
dtypes: float64(6), int64(2), object(4)
memory usage: 3.5+ MB


Check out if "TotalPayBenefits" = "BasePay" + "OvertimePay" + "OtherPay" + "Benefits"

In [15]:
salaries['TotalPayBenefitsFormula'] = salaries['BasePay'] + salaries['OvertimePay'] + salaries['OtherPay'] + salaries['Benefits']
salaries['TotalPayBenefitsFormula'].head()

0    510732.68
1    479652.21
2    436224.36
3    418019.22
4    417435.10
Name: TotalPayBenefitsFormula, dtype: float64

In [16]:
salaries['check'] = salaries['TotalPayBenefits'] == salaries['TotalPayBenefitsFormula']
salaries['check']

0         True
1        False
2         True
3        False
4         True
         ...  
38114     True
38115     True
38116     True
38117     True
38118     True
Name: check, Length: 38119, dtype: bool

What is the percetage of employees for which the previous assumption is not True?

In [17]:
salaries['check'].value_counts(normalize=True)*100

True     73.708649
False    26.291351
Name: check, dtype: float64

There are different departments in the city. List all departments and the count of employees in each department.

In [18]:
employee_by_dept = salaries.pivot_table(index = 'JobTitle', values= 'Id', aggfunc='count').rename({'Id':'Employee'})
employee_by_dept = employee_by_dept.rename(columns={'Id': 'Employees'})

# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of **all FT workers is significantly different from $75/hr**. Get first the hourly wage by dividing "TotalPayBenefits" by 50 weeks (assuming 10 labour days of holidays) and by 40hrs (assuming a 40hrs week).

$$Hourly Wage = \frac{TotalPayBenefits}{1 year}\frac{1 year}{50 Week}\frac{1 Week}{40 hr}$$

Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [19]:
# Computing the "Hourly_Wage"
salaries['Hourly_Wage'] = (salaries['TotalPayBenefits']/50)/40
salaries.head(40)

Id           EmployeeName                        JobTitle    BasePay  \
0   110532            David Shinn                  Deputy Chief 3  129150.01   
1   110533             Amy P Hart               Asst Med Examiner  318835.49   
2   110534   William J Coaker Jr.        Chief Investment Officer  257340.00   
3   110535         Gregory P Suhr                 Chief of Police  307450.04   
4   110536   Joanne M Hayes-White          Chief, Fire Department  302068.00   
5   110537        Ellen G Moffatt               Asst Med Examiner  270222.04   
6   110538          John L Martin                     Dept Head V  311298.55   
7   110539      Harlan L Kelly-Jr     Executive Contract Employee  310161.02   
8   110540             Samson Lai  Battalion Chief, Fire Suppress  179464.14   
9   110541       David L Franklin    Asst Chf of Dept (Fire Dept)  201566.88   
10  110542       Edward D Reiskin      Gen Mgr, Public Trnsp Dept  294000.18   
11  110543            Yifang Qian     Senior Physician Specialist  198874.01   
12  110544      Richard L Corriea                     Commander 3  184734.01   
13  110545       Raymond A Guzman     Dep Chf of Dept (Fire Dept)  260728.04   
14  110546            Jose L Velo  Battalion Chief, Fire Suppress  183333.82   
15  110547     Bryan W Rubenstein  Battalion Chief, Fire Suppress  177771.16   
16  110548         Roland Pickens              Administrator, DPH  282619.96   
17  110549        Mark A Gonzales     Dep Chf of Dept (Fire Dept)  260728.00   
18  110550       Barbara A Garcia                     Dept Head V  279839.22   
19  110551            Edwin M Lee                           Mayor  281537.17   
20  110552      Thomas A Siragusa    Asst Chf of Dept (Fire Dept)  207437.76   
21  110553       Denise A Schmitt                  Deputy Chief 3  268632.02   
22  110554            Lyn Tomioka                  Deputy Chief 3  268632.00   
23  110555        Arthur W Kenney    Asst Chf of Dept (Fire Dept)  207437.76   
24  110556         Greg B Pizarro               Asst Med Examiner  233386.01   
25  110557          Charlie Orkes                  Deputy Chief 3  265729.53   
26  110558           Susan Currin        Adm, SFGH Medical Center  268946.02   
27  110559          Naomi M Kelly                     Dept Head V  267914.01   
28  110560         Trent E Rhorer                     Dept Head V  267914.00   
29  110561  Benjamin L Rosenfield                      Controller  267914.00   
30  110562            Jay P Huish                     Dept Head V  267914.00   
31  110563          Kevin M Burke    Asst Chf of Dept (Fire Dept)  207437.76   
32  110564        Hector M Sainez                  Deputy Chief 3  258957.00   
33  110565           Marty A Ross  Battalion Chief, Fire Suppress  177771.08   
34  110566         Lorrie A Kalos  Battalion Chief, Fire Suppress  179464.20   
35  110567          Khairul A Ali  Battalion Chief, Fire Suppress  179464.21   
36  110568         Ai Kyung Chung                     Anesthetist  241036.02   
37  110569       Michael P Carlin                       Dep Dir V  254578.48   
38  110570        Sharon Ferrigno                  Deputy Chief 3  243339.02   
39  110571        Victor H Wyrsch  Battalion Chief, Fire Suppress  179464.20   

    OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0          0.00  342802.63  38780.04  471952.64         510732.68  2014   
1      10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2          0.00   82313.70  96570.66  339653.70         436224.36  2014   
3          0.00   19266.72  91302.46  326716.76         418019.22  2014   
4          0.00   24165.44  91201.66  326233.44         417435.10  2014   
5       6009.22   67956.20  71580.48  344187.46         415767.94  2014   
6          0.00       0.00  89772.32  311298.55         401070.87  2014   
7          0.00       0.00  88823.51  310161.02         398984.53  2014   
8     128685.99   27334.83  59876.90  335484.96     

In [20]:
# Computing the mean hourly wage for all the "FT" employees
ft_grouped = salaries[salaries['Status'] == 'FT']
ft_hourly_wage = ft_grouped['Hourly_Wage']
ft_hourly_wage_mean = ft_grouped['Hourly_Wage'].mean()
ft_hourly_wage_n = len(ft_hourly_wage)

### Hypothesis parameters and information

- H0: Hourly wage of all FT workers != 75 dolars/hr 
- H1: Hourly wage of all FT workers = 75 dolars/hr
- Confidence interval = 95%
- Alpha = 0.05
- Sample mean = 69.264206407271
- Sample size = 22334 (> 30)

In [21]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider "FT" employees
import scipy.stats

t_statistic_ft = (np.mean(ft_hourly_wage) - 75)/ (np.std(ft_hourly_wage, ddof=1)/np.sqrt(len(ft_hourly_wage)))
print('T_statistic is:', t_statistic_ft)

T_statistic is: -35.80631941460788


In [22]:
# Method 1: Critical value. Get the critical value and compare it against your statisttic.
critical_val_ft = scipy.stats.norm.ppf(1-(0.05/2))
print('Critical values are +', critical_val_ft,' and -', critical_val_ft)

Critical values are + 1.959963984540054  and - 1.959963984540054


Our t_statistic (-35.80) is much smaller than the critical value (-1.9599), so we can *reject* the null hypothesis than the hourly_wage of all FT workers is significantly different than 75 dolars per hour.

In [23]:
# Method 2: Use the p-value method.
# Your code here:

In [24]:
pval_ft = scipy.stats.norm.cdf(-35.80631941460788)
print('P-value is', pval_ft) #P-value super close to 0, we would REJECT the NULL HYPOTHESIS

P-value is 4.4031259419579245e-281


In [25]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:

from scipy.stats import ttest_1samp

# ttest_1samp(sample, population_mean)
stat, pval = ttest_1samp(ft_hourly_wage, popmean = 75, alternative = "two-sided")
print('T_statistic is:', stat)
print('P-value is:', pval)

T_statistic is: -35.80631941460526
P-value is: 4.517831233839448e-273


Are all the methods in agreement?

*Yes, the t-statistic suggests we should reject the null hypothesis, same as the p-value.*

## ----------------------------------------- ##

We are also curious about salaries in the police force. The chief of police in Chicago claimed in a press briefing that salaries this year are **higher than last year's mean of $86000/year for all salaried employees** (use the column "TotalPayBenefits". Test  hypothesis using a 95% confidence interval.

Hint: Use apply and a lambda function to check in "Police" is in the "JobTitle" to get all the "Police" jobs.

In [26]:
police_job = salaries['JobTitle'].apply(lambda x: 'Police' in x)
police_job_idx = police_job[police_job == True].index
police_salaries = salaries.loc[police_job_idx, 'TotalPayBenefits']
print(police_salaries.head())

3      418019.22
86     301749.81
143    289114.95
181    280222.76
300    267235.00
Name: TotalPayBenefits, dtype: float64


#### Some comments:
- H0: Police salaries are > than 86.000 dolars / year for all salaried employees
- H1: Police salaries are <= than 86.000 dolars / year for all salaried employees
- Confidence interval: 95%
- Alpha: 0.05
- Sample size = 1901 (>30)

In [27]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider "Police" employees

In [28]:
t_statistic_ps = (np.mean(police_salaries) - 86000)/ (np.std(police_salaries, ddof=1)/np.sqrt(len(police_salaries)))
print('T_statistic for police salaries hypothesis is:', t_statistic_ps)

T_statistic for police salaries hypothesis is: 50.252984742109916


In [29]:
# Method 1: Critical value. Get the critical value and compare it against your statisttic.
critical_val_ps = scipy.stats.norm.ppf(0.95)
print('Critical value is:', critical_val_ps)

Critical value is: 1.6448536269514722


In [30]:
# Method 2: Use the p-value method.

In [31]:
pval_ps = scipy.stats.norm.cdf(t_statistic_ps)
print('P-value is', pval_ps) 

P-value is 1.0


In [32]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy

In [33]:
stat, pval = ttest_1samp(police_salaries, popmean = 86000, alternative = "less")
print('T_statistic is:', stat)
print('P-value is:', pval)

T_statistic is: 50.252984742109945
P-value is: 1.0


*T_statistic is on the right side, very far from the critical value (rejection area on the left hand side). Also, the p-value is higher than 0.05, so both statistics indicate that we should ACCEPT the null hypothesis.*

## ---------------------------------------- ##

The workers from the "JobTitle" with the most employees have complained that their hourly wage is **less than $35/hour**. Using a one sample t-test, test this one-sided hypothesis at the 95% confidence level.

#### Comments:
- H0: transit operators hourly wage < 35 dolars / hour
- H1: transit operators hourly wage >= 35 dolars / hour
- Sample size: 2479 (>30)
- Sample mean: 43.35949773900775
- Confidence interval: 95%
- Alpha: 0.05

In [34]:
# Getting the department which has most employees
print('The department which has most employees is:', employee_by_dept.idxmax()[0])

The department which has most employees is: Transit Operator


In [35]:
transit_ops_hw = salaries[salaries['JobTitle'] == 'Transit Operator']
print('Sample size:', len(transit_ops_hw))
print('Sample mean:', transit_ops_hw['Hourly_Wage'].mean())
transit_ops_hw.head()

Sample size: 2479
Sample mean: 43.35949773900775


Id     EmployeeName          JobTitle   BasePay  OvertimePay  \
3223  113755    Nancy Padilla  Transit Operator  70636.56     78492.00   
3537  114069  Louis E Galarce  Transit Operator  71278.32     73221.72   
3766  114298      Duane Allen  Transit Operator  70029.28     69941.99   
4380  114912   Jaime O Olivar  Transit Operator  68842.56     65726.67   
4891  115423     Freeman Chee  Transit Operator  69777.22     60584.96   

      OtherPay  Benefits   TotalPay  TotalPayBenefits  Year         Agency  \
3223   6636.63  38481.57  155765.19         194246.76  2014  San Francisco   
3537   6785.53  38861.51  151285.57         190147.08  2014  San Francisco   
3766   8592.74  38733.41  148564.01         187297.42  2014  San Francisco   
4380   8828.06  38206.75  143397.29         181604.04  2014  San Francisco   
4891   8146.38  38480.23  138508.56         176988.79  2014  San Francisco   

     Status  TotalPayBenefitsFormula  check  Hourly_Wage  
3223     FT                194246.76   True    97.123380  
3537     FT                190147.08  False    95.073540  
3766     FT                187297.42   True    93.648710  
4380     FT                181604.04  False    90.802020  
4891     FT                176988.79   True    88.494395

In [36]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider the right "JobTitle" employees

In [37]:
t_statistic_to = (np.mean(transit_ops_hw['Hourly_Wage']) - 35)/ (np.std(transit_ops_hw['Hourly_Wage'], ddof=1)/np.sqrt(len(transit_ops_hw)))
print('T_statistic for transit operators hypothesis is:', t_statistic_to)

T_statistic for transit operators hypothesis is: 19.064990324906553


In [38]:
# Method 1: Critical value.
critical_val_to = scipy.stats.norm.ppf(0.95)
print('Critical value is:', critical_val_to)

Critical value is: 1.6448536269514722


In [39]:
# Method 2: Use the p-value method.
pval_to = scipy.stats.norm.cdf(t_statistic_to)
print('P-value is', pval_to) 

P-value is 1.0


In [40]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
stat, pval = ttest_1samp(transit_ops_hw['Hourly_Wage'], popmean = 35, alternative = "greater")
print('T_statistic is:', stat)
print('P-value is:', pval)

T_statistic is: 19.064990324906383
P-value is: 5.012771694204207e-76


*P-values don't coincide. Other than that, if we consider the t-statistics and the P-value of ttest_1samp, we would REJECT the null hypothesis (rejection area on the right hand side).*

# Challenge 3: To practice - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level.

In [41]:
# Sample mean
confidence_level = 0.95
hw_mean = salaries['Hourly_Wage'].mean()
sse = scipy.stats.sem(salaries['Hourly_Wage']) 
print('Hourly Wage mean for all workers:', round(hw_mean,2))
print('Sample standard error:', round(sse,2))

Hourly Wage mean for all workers: 50.13
Sample standard error: 0.17


In [42]:
confidence_interval = scipy.stats.norm.interval(confidence_level, 
                                             hw_mean, 
                                             sse)
print('Confidence interval:', confidence_interval)

Confidence interval: (49.79826596983879, 50.463172698540085)


Now compute a 95% confidence interval for the hourly salary of all the Police employees.

In [43]:
confidence_level_ps = 0.95
police_hw = salaries.loc[police_job_idx, 'Hourly_Wage']
police_hw_mean = police_hw.mean()
sse_ps = scipy.stats.sem(police_hw) 
print('Hourly Wage mean for police workers:', round(police_hw_mean,2))
print('Sample standard error:', round(sse_ps,2))

Hourly Wage mean for police workers: 74.17
Sample standard error: 0.62


In [44]:
confidence_interval_ps = scipy.stats.norm.interval(confidence_level_ps, 
                                             police_hw_mean, 
                                             sse_ps)
print('Confidence interval:', confidence_interval_ps)

Confidence interval: (72.9586542074945, 75.3903879177025)


# Chi2 test

Now we want to know if the amount of full time "FT" and part time "PT" employees is equal between Lawers, Meds, Police, Firemen and other departments. 

Considering all the options in this groups of employees will be very time consuming. To simplify this process, create first a function that returns:

* "Policemen" if "Police" is found on "JobTitle"
* "Firemen" if "Fire" is found on "JobTitle"
* "Medical" if "Med" or "Nurse" is found on "JobTitle"
* "Lawyer" if "Attorney" is found on "JobTitle"
* "Other" in any other cases

Then, create a new column named "employee_group" that determines to which group belong the employee. 

In [45]:
import scipy.stats as st

def group_jobtitles(x):
    if 'Police' in x:
        return 'Policemen'
    if 'Fire' in x:
        return 'Firemen'
    if 'Med' in x or 'Nurse' in x:
        return 'Medical'
    if 'Attorney' in x:
        return 'Lawyer'
    else:
        return 'Other'
    
salaries['employee_group'] = salaries['JobTitle'].apply(group_jobtitles)
salaries['employee_group'].value_counts()

Other        30371
Medical       3917
Policemen     1901
Firemen       1511
Lawyer         419
Name: employee_group, dtype: int64

Determine how many "PT" and "FT" employess have all the employees groups.

In [46]:
# Your code here: (Store the output dataframe into a new variable)
freq_status_employee = pd.crosstab(salaries['employee_group'], salaries['Status'])
display(freq_status_employee)

Status             FT     PT
employee_group              
Firemen          1333    178
Lawyer            317    102
Medical          1028   2889
Other           18126  12245
Policemen        1530    371

Now try compute the expected frequencies doing the calculations with the individual probabilities. Remember that the Chi2 test assumes that both variables (employee_group and FT/PT) are not related (therefore they are independent). Therefore, to compute the expected frequencies you need to compute the probability of each cell and multiply it by the number of observations. ie:

$$\nu(x,y) = p(x,y) * N = p(x) * p(y) * N$$

bear in mind that in general: $p(x,y)\neq p(x)*p(y)$; the equality will only be true if x and y are independent. However, the null hypotheses says that **x and y are independent.** but that's what we're assuming with the null hypotheses.

where "x" is the "employee_group" and "y" the (FT/PT). 

In [47]:
# Create an empty dataframe named "frequencies" to store the data.
frequencies = pd.DataFrame()
print(type(frequencies))

<class 'pandas.core.frame.DataFrame'>


In [48]:
# Compute Expected frequency of being "Firemen" and "FT". Store the solution in a variable named "firemen_ft"

In [49]:
P_firemen = freq_status_employee.values[0].sum()/freq_status_employee.values.sum()
print('Probability of being a firemen:', round(P_firemen,2))
P_FT = freq_status_employee['FT'].sum() / freq_status_employee.values.sum()
print('Probability of being a FT worker:', round(P_FT,2))

firemen_ft = (P_firemen * P_FT) * freq_status_employee.values.sum()
print(round(firemen_ft,2))

Probability of being a firemen: 0.04
Probability of being a FT worker: 0.59
885.3


In [50]:
# Compute Expected frequency of being "Firemen" and "PT". Store the solution in a variable named "firemen_pt"

In [51]:
P_PT = freq_status_employee['PT'].sum() / freq_status_employee.values.sum()
print('Probability of being a firemen:', round(P_firemen,2))
print('Probability of being a PT worker:', round(P_PT,2))

firemen_pt = (P_firemen * P_PT) * freq_status_employee.values.sum()
print(round(firemen_pt,2))

Probability of being a firemen: 0.04
Probability of being a PT worker: 0.41
625.7


In [52]:
# Compute Expected frequency of being "Lawyers" and "FT". Store the solution in a variable named "lawyers_ft"

In [53]:
P_lawyer = freq_status_employee.values[1].sum()/freq_status_employee.values.sum()
print('Probability of being a lawyer:', round(P_lawyer,2))
print('Probability of being a FT worker:', round(P_FT,2))

lawyers_ft = (P_lawyer * P_FT) * freq_status_employee.values.sum()
print(round(lawyers_ft,2))

Probability of being a lawyer: 0.01
Probability of being a FT worker: 0.59
245.49


In [54]:
# Compute Expected frequency of being "Lawyers" and "PT". Store the solution in a variable named "lawyers_pt"

In [55]:
print('Probability of being a lawyer:', round(P_lawyer,2))
print('Probability of being a PT worker:', round(P_PT,2))

lawyers_pt = (P_lawyer * P_PT) * freq_status_employee.values.sum()
print(round(lawyers_pt,2))

Probability of being a lawyer: 0.01
Probability of being a PT worker: 0.41
173.51


In [56]:
# Compute Expected frequency of being "Medical" and "FT". Store the solution in a variable named "medical_ft"

In [57]:
P_medical = freq_status_employee.values[2].sum()/freq_status_employee.values.sum()
print('Probability of being a medical:', round(P_medical,2))
print('Probability of being a FT worker:', round(P_FT,2))

medical_ft = (P_medical * P_FT) * freq_status_employee.values.sum()
print(round(medical_ft,2))

Probability of being a medical: 0.1
Probability of being a FT worker: 0.59
2294.98


In [58]:
# Compute Expected frequency of being "Medical" and "PT". Store the solution in a variable named "medical_pt"

In [59]:
print('Probability of being a medical:', round(P_medical,2))
print('Probability of being a PT worker:', round(P_PT,2))

medical_pt = (P_medical * P_PT) * freq_status_employee.values.sum()
print(round(medical_pt,2))

Probability of being a medical: 0.1
Probability of being a PT worker: 0.41
1622.02


In [60]:
# Compute Expected frequency of being "Other" and "FT". Store the solution in a variable named "other_ft"

In [61]:
P_other = freq_status_employee.values[3].sum()/freq_status_employee.values.sum()
print('Probability of being others:', round(P_other,2))
print('Probability of being a FT worker:', round(P_FT,2))

other_ft = (P_other * P_FT) * freq_status_employee.values.sum()
print(round(other_ft,2))

Probability of being others: 0.8
Probability of being a FT worker: 0.59
17794.43


In [62]:
# Compute Expected frequency of being "Other" and "PT". Store the solution in a variable named "other_pt"

In [63]:
print('Probability of being others:', round(P_other,2))
print('Probability of being a PT worker:', round(P_PT,2))

other_pt = (P_other * P_PT) * freq_status_employee.values.sum()
print(round(other_pt,2))

Probability of being others: 0.8
Probability of being a PT worker: 0.41
12576.57


In [64]:
# Compute Expected frequency of being "Policement" and "FT". Store the solution in a variable named "policemen_ft"

In [65]:
P_policemen = freq_status_employee.values[4].sum()/freq_status_employee.values.sum()
print('Probability of being a policeman:', round(P_policemen,2))
print('Probability of being a FT worker:', round(P_FT,2))

policemen_ft = (P_policemen * P_FT) * freq_status_employee.values.sum()
print(round(policemen_ft,2))

Probability of being a policeman: 0.05
Probability of being a FT worker: 0.59
1113.8


In [66]:
# Compute Expected frequency of being "Policement" and "PT". Store the solution in a variable named "policemen_pt"

In [67]:
print('Probability of being a policeman:', round(P_policemen,2))
print('Probability of being a PT worker:', round(P_PT,2))

policemen_pt = (P_policemen * P_PT) * freq_status_employee.values.sum()
print(round(policemen_pt,2))

Probability of being a policeman: 0.05
Probability of being a PT worker: 0.41
787.2


* Store all the expected frequencies of "FT" employees in a list 
* Store all the "PT" employees into another list
* Create a dictionary with "FT" and "PT" as keys and as the values use the previous lists
* Create a dataframe with this dictionary using pd.DataFrame()

In [68]:
FT_lst = [firemen_ft, lawyers_ft, medical_ft, other_ft, policemen_ft]
PT_lst = [firemen_pt, lawyers_pt, medical_pt, other_pt, policemen_pt]

print('Full Time employees:', FT_lst)
print('Part Time employees:', PT_lst)

Full Time employees: [885.2979878800596, 245.49295626852748, 2294.9783047823917, 17794.430966184842, 1113.7997848841785]
Part Time employees: [625.7020121199403, 173.5070437314725, 1622.021695217608, 12576.569033815158, 787.2002151158215]


In [69]:
freq_expected_dict = {'FT': FT_lst,
            'PT': PT_lst}
freq_expected_df = pd.DataFrame(freq_expected_dict, index = freq_status_employee.index)
display(freq_expected_df)

FT            PT
employee_group                            
Firemen           885.297988    625.702012
Lawyer            245.492956    173.507044
Medical          2294.978305   1622.021695
Other           17794.430966  12576.569034
Policemen        1113.799785    787.200215

Now use the "st.chi2_contingency()" from scipy.stats [documentation here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html) to conduct a Chi2 test to determine if the diferences between employee groups are statistically significant using a 95% confidence level. Hint: fill the function with a dataframe of actual frequencies.

In [70]:
# Your code here: (use the st.chi2_contingency() function from scipy.stats to compute:
# The Chi2 value
# The p-value 
# The expected frequencies.

#### Comments:
- Ho: x and y are independent
- Ha: x and y are NOT independent

In [71]:
chi2_test = st.chi2_contingency(np.array(freq_status_employee))

print(chi2_test)

(2676.642333711905, 0.0, 4, array([[  885.29798788,   625.70201212],
       [  245.49295627,   173.50704373],
       [ 2294.97830478,  1622.02169522],
       [17794.43096618, 12576.56903382],
       [ 1113.79978488,   787.20021512]]))


In [72]:
print('Chi2 value:', chi2_test[0])
print('P-value:', chi2_test[1])
print('Degrees of freedom:', chi2_test[2])
print('Expected frequencies array:\n', chi2_test[3])
chi_cv = st.chi2.ppf(0.95,df=chi2_test[2])
print('Critical value:', chi_cv)

Chi2 value: 2676.642333711905
P-value: 0.0
Degrees of freedom: 4
Expected frequencies array:
 [[  885.29798788   625.70201212]
 [  245.49295627   173.50704373]
 [ 2294.97830478  1622.02169522]
 [17794.43096618 12576.56903382]
 [ 1113.79978488   787.20021512]]
Critical value: 9.487729036781154


*As chi value 2676.642333711905 > 9.487729036781154 and the p-value is smaller than 0.05, we REJECT the null hypothesis and we can conclude that the features are NOT INDEPENDENT. Therefore, the features are somehow related.*

Check if your expected frequencies aggree with the ones obtained with the st.chi2_contingency() function

In [73]:
chi_freq = pd.DataFrame(chi2_test[3], index = freq_expected_df.index, columns = freq_expected_df.columns )
display(chi_freq)
display(freq_expected_df)

FT            PT
employee_group                            
Firemen           885.297988    625.702012
Lawyer            245.492956    173.507044
Medical          2294.978305   1622.021695
Other           17794.430966  12576.569034
Policemen        1113.799785    787.200215

FT            PT
employee_group                            
Firemen           885.297988    625.702012
Lawyer            245.492956    173.507044
Medical          2294.978305   1622.021695
Other           17794.430966  12576.569034
Policemen        1113.799785    787.200215